src/librbd/operation/DisableFeaturesRequest.h

```diff
@@ -0,0 +1,162 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_OPERATION_DISABLE_FEATURES_REQUEST_H
+#define CEPH_LIBRBD_OPERATION_DISABLE_FEATURES_REQUEST_H
+
+#include "librbd/operation/Request.h"
+#include "cls/rbd/cls_rbd_client.h"
+
+class Context;
+
+namespace librbd {
+
+class ImageCtx;
+
+namespace operation {
+
+template <typename ImageCtxT = ImageCtx>
+class DisableFeaturesRequest : public Request<ImageCtxT> {
+public:
+  static DisableFeaturesRequest *create(ImageCtxT &image_ctx, Context *on_finish,
+                                        uint64_t journal_op_tid,
+                                        uint64_t features) {
+    return new DisableFeaturesRequest(image_ctx, on_finish, journal_op_tid,
+                                      features);
+  }
+
+  DisableFeaturesRequest(ImageCtxT &image_ctx, Context *on_finish,
+                         uint64_t journal_op_tid, uint64_t features);
+
+protected:
+  virtual void send_op();
+  virtual bool should_complete(int r);
+  virtual bool can_affect_io() const override {
+    return true;
+  }
+  virtual journal::Event create_event(uint64_t op_tid) const {
+    return journal::UpdateFeaturesEvent(op_tid, m_features, false);
+  }
+
+private:
+  /**
+   * DisableFeatures goes through the following state machine:
+   *
+   * @verbatim
+   *
+   * <start>
+   *    |
+   *    v
+   * STATE_BLOCK_WRITES
+   *    |
+   *    v
+   * STATE_ACQUIRE_EXCLUSIVE_LOCK (skip if not
+   *    |                          required)
+   *    | (disbling journaling)
+   *    \-------------------\
+   *    |                    |
+   *    |                    V
+   *    |                 STATE_GET_MIRROR_MODE
+   *    |(not                |
+   *    | disabling          v
+   *    | journaling)     STATE_GET_MIRROR_IMAGE
+   *    |                    |
+   *    |                    v
+   *    |                 STATE_DISABLE_MIRROR_IMAGE (skip if not
+   *    |                    |                        required)
+   *    |                    v
+   *    |                 STATE_CLOSE_JOURNAL
+   *    |                    |
+   *    |                    v
+   *    |                 STATE_REMOVE_JOURNAL
+   *    |                    |
+   *    |/-------------------/
+   *    |
+   *    v
+   * STATE_REMOVE_OBJECT_MAP (skip if not
+   *    |                     disabling object map)
+   *    v
+   * STATE_SET_FEATURES
+   *    |
+   *    v
+   * STATE_UPDATE_FLAGS
+   *    |
+   *    v
+   * STATE_APPEND_OP_EVENT (skip if journaling
```

Comment:
I think this needs to be before REMOVE_OBJECT_MAP since that would be the first modification state